DIP-10 Updates (#164)
diem/dip: dips/dip-10.md

* VASP A contacts VASP B via an off-chain API with the sender identifier `alice@avasp` and requests a reference_id, `rb`
* VASP A constructs a transaction with the specified amount and the reference_id `rb` and submits it to the Diem network
* VASP B receives a transaction with metadata containing `rb`, deposits the amount in Bob's account. (VASP B may use Alice’s DiemID for its own record keeping or seek receipt confirmation from Bob.)

# DiemID Format
Diem defines globally unique identifiers to facilitate transacting on the Diem Payment Network. As Diem stores no personally identifiable information on-chain, these identifiers are exchanged during the off-chain process between two VASPs to identify a distinct source and destinations within their respective VASPs in order to produce a reference_id that can be stored on-chain. The format of identifiers is `[user_identifier]@[vasp_domain_identifier]`

Example: `alice@avasp`

* `user_identifier` is a reusable identifier that represents either the source or destination end-user of a payment transaction. It is unique to per user at VASP level. Specification:
  * Case insensitive

>   * Valid regular expression: `^[a-zA-Z0-9][a-zA-Z0-9.]*$`

Reviewer: it might be worthwhile having a section at the bottom describing how we settled on this, how compatible is it with other naming schemes?

# Summary

This DIP describes DiemID - the human-readable identifier for user accounts.   

# Motivation
DiemID provides a convenient method for identifying users within a VASP. DiemID allows users to exchange human-readable identifiers as either the sender or the receiver of peer-to-peer payments, and plays the role of an email address for payments. The benefits of using a DiemID are:

> * Privacy: DiemID needs to be used in tandem with an off-chain protocol between VASPs in order to exchange information about the end user corresponding to the DiemID. By initializing an off-chain preflight to exchange a reference ID, we do not need to include any potentially identifiable user information in a transaction 

Reviewer: ```suggestion
* Privacy: DiemID needs to be used in tandem with an off-chain protocol between VASPs in order to exchange information about the end user corresponding to the DiemID. By initializing an off-chain preflight reference ID exchange, DiemID does require any potentially identifiable user information in an on-chain transaction 
```

> * Persistent Identifiers: Currently there is a lack of persistent user identifiers in the Diem ecosystem. DiemID is a persistent identifier from a user's perspective. From the perspective of the chain, DiemIDs do not exist

Reviewer: ```suggestion
* Persistent Identifiers: Currently there is a lack of persistent user identifiers in the Diem ecosystem. DiemID is a persistent identifier from a user's perspective. From the perspective of the on-chain transactions, DiemIDs are irrelevant and provide no meaningful information
```

# End-to-End Experience
Below is an example of using DiemID domain for transferring money from one user to another. 

## Prerequisite:

> * VASPs get approval from association (via some offline process) on domain name. VASPs optionally have locally stored a mapping of domain ID to parent VASP address by reading the DiemID event stream under Treasury Compliance on-chain account:

Reviewer: Maybe make two bullets here

```suggestion
* VASPs get approval from association (via some offline process) on domain name.
* VASPs have access to a storage solution mapping  domain IDs to parent VASP address
```

the rest are implementation details discussed below

```
avasp, 0xf72589b71ff4f8d139674a3f7369c69b
bvasp, 0xc5ab123458df0003415689adbb47326d
```

> ## Illustration - An example P2P Transaction Flow: 

Reviewer: ```suggestion
An Example P2P Transaction Flow: 
```

* Bob wants to receive funds from Alice
* Bob registers a DiemID at VASP B: `bob@bvasp`
* Alice's DiemID is registered at VASP A: `alice@avasp`
* Bob shares the DiemID, `bob@bvasp` with Alice
* Alice logs into VASP A, enters Bob’s identifier, an amount to pay, and submits payment

> * Alice’s VASP (VASP A) contacts Bob’s VASP’s (VASP B) off-chain API with the sender identifier `alice@avasp` and requests a reference_id, `rb`

Author: This is so that Bob can know where the payment came from, because the onchain transaction won't have any sending user information

> * `vasp_domain_identifier` is a unique string that is mapped to a VASP. Specification:

Reviewer: @aylinaykol / @sunmilee I think we should definitely consider a "." here so that we can support domains and not be completely disjoint from paystring (food for thought)

Author: okie i agree i added .

Author: Should it be `^[a-zA-Z0-9][a-zA-Z0-9.]*$` or `[a-zA-Z0-9.]+`  ?

  * Only TC account can create and publish `DiemIdDomains` resource
  * Only TC account can add, remove or update an `DiemIdDomain` within `DiemIdDomains` resource
* `DiemIDDomains` resource will be created in an empty state as part of creating a `ParentVASP` account resource, and existing `ParentVASP` accounts without `DiemIDDomains` will have the resource instantiated by the DiemRoot account.
* In order to register a DiemID domain, a VASP needs to submit a request to Diem Association. Diem Association will perform certain checks (out of scope of this document) before issuing an on-chain transaction to register a DiemID Domain. These checks intend to mitigate irrelevant claims and enforce uniqueness

## DiemID Domain Events

The Move contract that manages DiemID domains must emit an event every time DiemID domain is created, removed or updated. Those events are critical for applications to be able to efficiently index existing DiemID domains.
An application can be built to listen for these events to construct a mapping of DiemID domains to VASP accounts for lookup of onchain addresses given a DiemID domain.

> While DiemID domains are published into VASP account resource, DiemID domain events are published under the Treasury Compliance account. We consolidate events under single account to allow indexers to follow a single event stream.

Reviewer 2: Were you able to consult with the move team on this mechanism?

Author: Yeah move team has reviewed the last PR

}
```
* Fields definition:
   * `domain` - name of a DiemID domain 
* The `DiemIdDomains` resource can only be published into an account with `Role == PARENT_VASP_ROLE_ID`.
* The `DiemIdDomains` contains a list of `DiemIdDomain` structs that are associated with a VASP. As such, it is possible to register more than one DiemID Domain for a given VASP
* Only special Treasury Compliance account (address `0xB1E55ED`) can manipulate DiemIdDomains resource:
  * Only TC account can create and publish `DiemIdDomains` resource
  * Only TC account can add, remove or update an `DiemIdDomain` within `DiemIdDomains` resource
* `DiemIDDomains` resource will be created in an empty state as part of creating a `ParentVASP` account resource, and existing `ParentVASP` accounts without `DiemIDDomains` will have the resource instantiated by the DiemRoot account.

> * In order to register a DiemID domain, a VASP needs to submit a request to Diem Association. Diem Association will perform certain checks (out of scope of this document) before issuing an on-chain transaction to register a DiemID Domain. These checks intend to mitigate irrelevant claims and enforce uniqueness

Reviewer 3: We should avoid concrete governance principles of the Diem Group here. DIPs should refer to on-chain roles/permissions, referencing DIP-2 as needed. See example in DIP-20.

## DiemID Domain Events

The Move contract that manages DiemID domains must emit an event every time DiemID domain is created, removed or updated. Those events are critical for applications to be able to efficiently index existing DiemID domains.
An application can be built to listen for these events to construct a mapping of DiemID domains to VASP accounts for lookup of onchain addresses given a DiemID domain.

> While DiemID domains are published into VASP account resource, DiemID domain events are published under the Treasury Compliance account. We consolidate events under single account to allow indexers to follow a single event stream.

Reviewer 3: ```suggestion
While DiemID domains are published into VASP account resource, DiemID domain events are published under the TreasuryCompliance account. We consolidate events under single account to allow indexers to follow a single event stream.
```

> To store events, `DiemIdDomainManager` resource is published under the Treasure Compliance account(address `0xB1E55ED`).

Reviewer 3: ```suggestion
To store events, `DiemIdDomainManager` resource is published under the TreasuryCompliance account(address `0xB1E55ED`).
```

# ReferenceID Exchange

> The PaymentMetadata transactions require a ReferenceID in order to submit a transaction to the chain. PaymentMetadata transactions reveal nothing distinctly about either the sender or receiver and do not create a linkability between the sender or receiver across payments. Hence, the ReferenceID must be established off-chain and this protocol defines one such way to do so leveraging DiemID.

Reviewer 3: ```suggestion
PaymentMetadata transactions require a ReferenceID in order to submit a transaction to the chain. PaymentMetadata transactions reveal nothing distinctly about either the sender or receiver and do not create a linkability between the sender or receiver across payments. Hence, the ReferenceID must be established off-chain and this protocol defines one such way to do so leveraging DiemID.
```

Reviewer 3: "PaymentMetadata transaction" reads wrong to me. Isn't PaymentMetadata a field, not a transaction?

> A VASP intending to send a payment from one VASP to another when leveraging DiemIDs constructs a ReferenceIDCommand submits that to the receiver. If the receiver knows the recipient DiemID and can potentially accept payments, it returns a success. In the case it cannot, it returns an error indicating the reason.

Reviewer 3: ```suggestion
A VASP intending to send a payment to another using DiemIDs constructs a ReferenceIDCommand and transmits it to the receiver. If the receiver knows the recipient DiemID and can potentially accept payments, it returns a success. Otherwise, it returns an error indicating the reason.
```

| Field 	    | Type 	     | Required? 	| Description 	           |
|-------	    |------	     |-----------	|-------------	           |
| _ObjectType   | str    | Y | Fixed value: `ReferenceIDCommand`|
| sender        | str          | Y            | Sender's full DiemID |

> | sender_address| str          | Y            | Sender's on-chain address |

Author: Could be interesting maybe i can bring it to tech review as a discussion point

> | sender_address| str          | Y            | Sender's on-chain address |

Reviewer: would love to do two things:
embrace bech32
add definition for bech32 without subaddress
and well a third is to get rid of specialized bech32 implementations

> | sender_address| str          | Y            | Sender's on-chain address |

Reviewer: maybe we just add bech32 without subaddress here? since dip-5 is already finalized

> | sender_address| str          | Y            | Sender's on-chain address |

Author: i think we should have a separate dip for bech32 generic forms

> | sender_address| str          | Y            | Sender's on-chain address |

Author: Should we still call it "sender_address" and ?receiver_address"? @davidiw @xli

> | sender_address| str          | Y            | Sender's [bech32 identifier](https://github.com/diem/dip/blob/main/dips/dip-5.md) on-chain address |

Reviewer 4: May consider use name "account identifier" instead of bech32 identifier, because it is defined in dip-5. Also need to mention subaddress should be None / 00000000

    "status": "success",
    "result": {
	    "_ObjectType": "ReferenceIDCommandResponse",

> 	    "receiver_address": "c5ab123458df0003415689adbb47326d",

Author: Yeah it should be consistent to either both bech32 or neither -- just need to decide if we want bech32 or just write chain ID as a separate field?

    },
    "cid": "12ce83f6-6d18-0d6e-08b6-c00fdbbf085a",
}
```

The format of the failed response is:
```
{
   "_ObjectType": "CommandResponseObject",
    "status": "failure",

>     "result": {

Author: makes sense

    "status": "success",
    "result": {
	    "_ObjectType": "ReferenceIDCommandResponse",
	    "receiver_address": "dm1p7ujcndcl7nudzwt8fglhx6wxn08kgs5tm6mz4us2vfufk",
    },
    "cid": "12ce83f6-6d18-0d6e-08b6-c00fdbbf085a",
}
```

> The format of the failed response is:

Reviewer: why? this is defined in dip1?

```
{
   "_ObjectType": "CommandResponseObject",
    "status": "failure",
    "cid": "12ce83f6-6d18-0d6e-08b6-c00fdbbf085a",
    "error": OffChainErrorObject(),
}
```

> **CommandResponseObject:**

Reviewer: why? this is defined in dip1 and your latest dip?

| Field 	    | Type 	     | Required? 	| Description 	           |
|-------	    |------	     |-----------	|-------------	           |
| _ObjectType   | str        | Y | Fixed value: `CommandResponseObject`|
| status       | str | Y | Either `success` or `failure`. |
| result       | Result object | N | The Result obejct of response. |
| cid           | str         | Y            | A unique identifier for the Command. Should be a UUID according to [RFC4122](https://tools.ietf.org/html/rfc4122) with "-"'s included. |
| error          | [OffChainErrorObject](https://github.com/diem/dip/blob/main/dips/dip-1.mdx#offchainerrorobject) | N | Details of the error when status == "failure" |

**CommandResultObject:**

| Field 	    | Type 	     | Required? 	| Description 	           |
|-------	    |------	     |-----------	|-------------	           |
| _ObjectType   | str        | Y | Fixed value: `ReferenceIDCommandResponse`|

> | receiver_address       | str | Y | Receiver's onchain [account identifier](https://github.com/diem/dip/blob/main/dips/dip-5.md) with subaddress set to `None` or `00000000`|

Reviewer 4: my bad, according to dip-5, may use ``` `0x0000000000000000` ``` instead of ``` `None` or `00000000` ```

Reviewer: 8 bytes means 16 hex values

Reviewer: I would just say `None` or the zero subaddress?

> #### Error Codes

Reviewer: can use invalid_field_value for invalid fields but we should be explicit what the cases are (I think that would have actually smoothened the convo in the preso today)

> `duplicate_ref_id`: Duplicate Reference ID was rejected by the receiving end

Reviewer: ```suggestion
`duplicate_reference_id`: Duplicate Reference ID was rejected by the receiving end
```

| status       | str | Y | Either `success` or `failure`. |

> | result       | Result object | Y | The Result obejct of response. |

Reviewer 4: Result is optional

| cid           | str         | Y            | A unique identifier for the Command. Should be a UUID according to [RFC4122](https://tools.ietf.org/html/rfc4122) with "-"'s included. |
| error          | [OffChainErrorObject](https://github.com/diem/dip/blob/main/dips/dip-1.mdx#offchainerrorobject) | N | Details of the error when status == "failure" |

**CommandResultObject:**

| Field 	    | Type 	     | Required? 	| Description 	           |
|-------	    |------	     |-----------	|-------------	           |
| _ObjectType   | str        | Y | Fixed value: `ReferenceIDCommandResponse`|

> | receiver_address       | str | Y | Receiver's [bech32 identifier](https://github.com/diem/dip/blob/main/dips/dip-5.md) |

Reviewer 4: May consider use "account identifier" instead of bech32 identifier, because it is defined in dip-5

| status       | str | Y | Either `success` or `failure`. |
| result       | Result object | Y | The Result obejct of response. |
| cid           | str         | Y            | A unique identifier for the Command. Should be a UUID according to [RFC4122](https://tools.ietf.org/html/rfc4122) with "-"'s included. |
| error          | [OffChainErrorObject](https://github.com/diem/dip/blob/main/dips/dip-1.mdx#offchainerrorobject) | N | Details of the error when status == "failure" |

**CommandResultObject:**

| Field 	    | Type 	     | Required? 	| Description 	           |
|-------	    |------	     |-----------	|-------------	           |
| _ObjectType   | str        | Y | Fixed value: `ReferenceIDCommandResponse`|
| receiver_address       | str | Y | Receiver's on-chain address |

#### Error Codes

`duplicate_ref_id`: Duplicate Reference ID was rejected by the receiving end

> `invalid_user`: Receiving end could not find the user with the given user_identifier

Reviewer: should we distinguish `invalid_receiver` and `unknown_sender` and `unknown_sender_address` and `invalid_sender_address` for when the sender address isn't the same as the txn sender or one of its child accounts?

Author: How would invalid_sender_address get sent back? the reference ID exchange would've ended already by the time the receiver has gotten the on-chain transaction?

type ReferenceId = [u8, 16];
```

> If the amount is at or above the travel rule limit, VASPs should trigger an off-chain protocol for KYC exchange. The same `reference_id` must be used to perform a TR as described in [DIP-1](https://github.com/diem/dip/blob/master/dips/dip-1.md).

Reviewer: ```suggestion
If the amount is at or above the travel rule limit, VASPs should trigger an off-chain protocol for KYC exchange. The same `reference_id` must be used to perform a travel rule exchange as described in [DIP-1](https://github.com/diem/dip/blob/master/dips/dip-1.md).
```

Author: good catch!

```

> If the amount exceeds the travel rule limit, VASPs should trigger an off-chain protocol for KYC exchange. The same `reference_id` must be used to perform a TR as described in [DIP-1](https://github.com/diem/dip/blob/master/dips/dip-1.md).

Reviewer 3: exceeds --> at or above (the travel rule limit)
